src/core/reactor.cc

```diff
@@ -2738,6 +2745,29 @@ reactor::insert_activating_task_queues() {
     _activating_task_queues.clear();
 }
 
+void reactor::add_task(task* t) noexcept {
+    auto sg = t->group();
+    auto* q = _task_queues[sg._id].get();
+    bool was_empty = q->_q.empty();
+    q->_q.push_back(std::move(t));
+    shuffle(q->_q.back(), q->_q);
+    if (was_empty) {
+        activate(*q);
+    }
+}
+
+void reactor::add_urgent_task(task* t) noexcept {
+    memory::scoped_critical_alloc_section _;
+    auto sg = t->group();
+    auto* q = _task_queues[sg._id].get();
+    bool was_empty = q->_q.empty();
+    q->_q.push_front(std::move(t));
+    shuffle(q->_q.front(), q->_q);
+    if (was_empty) {
+        activate(*q);
+    }
```

Comment:
```suggestion
    add_task(t);
```

it seems the body guarded by `scoped_critical_alloc_section` is identical to `add_task()`. so probably we can reuse it?

Reply:
as it's not in the scope of this commit. might want to do this in a dedicated commit in this PR.

Reply:
No, the bodies are different. The "_urgent_" version pushes task to front and shufflet front element, the regular one pushes to back and shuffle back element